runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1281 @@
+	pub enum Event<T: Config> {
+		/// A para was assigned a permanent parachain slot
+		PermanentSlotAssigned(ParaId),
+		/// A para was assigned a temporary parachain slot
+		TemporarySlotAssigned(ParaId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The specified parachain or parathread is not registered.
+		ParaDoesntExist,
+		/// Not a parathread.
+		NotParathread,
+		/// Cannot upgrade parathread.
+		CannotUpgrade,
+		/// Cannot downgrade parachain.
+		CannotDowngrade,
+		/// Permanent or Temporary slot already assigned.
+		SlotAlreadyAssigned,
+		/// Permanent or Temporary slot has not been assigned.
+		SlotNotAssigned,
+		/// An ongoing lease already exists.
+		OngoingLeaseExists,
+		// Maximum number of permanent slots exceeded
+		MaxPermanentSlotsExceeded,
+		// Maximum number of temporary slots exceeded
+		MaxTemporarySlotsExceeded,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			if let Some((lease_period, first_block)) = Self::lease_period_index(n) {
+				// If we're beginning a new lease period then handle that.
+				if first_block {
+					return Self::manage_lease_period_start(lease_period)
+				}
+			}
+
+			// We didn't return early above, so we didn't do anything.
+			0
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Assign a permanent parachain slot and immediately create a lease for it.
+		#[pallet::weight((1_000 + <T as slots::Config>::WeightInfo::force_lease(), DispatchClass::Operational))]
+		pub fn assign_perm_parachain_slot(origin: OriginFor<T>, id: ParaId) -> DispatchResult {
+			T::AssignSlotOrigin::ensure_origin(origin)?;
+
+			let manager = T::Registrar::manager_of(id).ok_or(Error::<T>::ParaDoesntExist)?;
+
+			ensure!(T::Registrar::is_parathread(id), Error::<T>::NotParathread,);
+
+			ensure!(
+				!Self::has_permanent_slot(id) && !Self::has_temporary_slot(id),
+				Error::<T>::SlotAlreadyAssigned
+			);
+
+			let current_lease_period: T::BlockNumber = Self::current_lease_period_index();
+			ensure!(
+				!T::Leaser::already_leased(
+					id,
+					current_lease_period,
+					// Check current lease & next one
+					current_lease_period.saturating_add(
+						T::BlockNumber::from(2u32)
+							.saturating_mul(T::PermanentSlotLeasePeriodLength::get().into())
+					)
+				),
+				Error::<T>::OngoingLeaseExists
+			);
+
+			ensure!(
+				PermanentSlotCount::<T>::get() < T::MaxPermanentSlots::get(),
+				Error::<T>::MaxPermanentSlotsExceeded
+			);
+
+			<PermanentSlotCount<T>>::try_mutate(|count| -> DispatchResult {
+				// Permanent slot assignment fails if a lease cannot be created
+				Self::configure_slot_lease(
+					id,
+					manager,
+					current_lease_period,
+					T::PermanentSlotLeasePeriodLength::get().into(),
+				)
+				.map_err(|_| Error::<T>::CannotUpgrade)?;
+				*count = count.checked_add(One::one()).ok_or(ArithmeticError::Overflow)?;
```

Comment:
configure_slot_lease appears to write to storage, thus you are not allowed to fail here on the `checked_add`.

Either do the checked add before the storage write, or use saturating, or anything but return an error after a storage write.